src/compile.c

```diff
@@ -375,8 +378,32 @@ static int block_bind_subblock_inner(int* any_unbound, block binder, block body,
     // binding recurses into argument list
     nrefs += block_bind_subblock_inner(&i->any_unbound, binder, i->arglist, bindflags, break_distance);
 
-    if (i->any_unbound)
+    if (i->any_unbound) {
       *any_unbound = 1;
+      if (!fast_path) {
+        /*
+         * This inst is unbound in the slow path, so we hook it up into a
+         * sub-tree of unbound insts.
+         *
+         * Note that we're going backwards from body.last to body.first in the
+         * slow path.
+         */
+        i->next_unbound = prev;
+      }
+      prev = i;
+    } else if (!fast_path && prev) {
+      /*
+       * This inst is bound, so it need not go into the unbounds sub-tree, but
+       * we make its next_unbound point into the unbounds sub-tree anyways.
+       */
+      i->next_unbound = prev; /* ditto */
+    } else if (fast_path && i != body.first) {
+      /*
+       * In this case we must have finally fully-bound the sub-tree at this
+       * inst, so we remove this inst from the sub-tree of unbounds.
+       */
+      prev->next_unbound = i->next_unbound;
+    }
```

Comment:
If we add `else if (fast_path) printf("\n");` here we see 5528 cases in the first 154 tests in `src/jq.test` where we don't benefit because `i == body.first` and we can't remove `body.first` from the unbound sub-tree.